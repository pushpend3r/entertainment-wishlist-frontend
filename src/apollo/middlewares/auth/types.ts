import { Operation } from "@apollo/client";
import { Tokens } from "../../../gql/graphql";

export interface FailedOperation {
  operation: Operation;
  settlePromiseWith: (value: Tokens | null) => void;
}
