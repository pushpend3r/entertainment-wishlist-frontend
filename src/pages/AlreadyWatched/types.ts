import { Movie, TVShow } from "../../types";

export interface GetAlreadyWatchedMedia {
  getWatchedMovies: Movie[];
  getWatchedTVShows: TVShow[];
}
