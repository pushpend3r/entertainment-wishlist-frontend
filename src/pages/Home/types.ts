import { Movie, TVShow } from "../../types";

export interface GetTrending {
  trending: {
    movies: Movie[];
    tvshows: TVShow[];
  };
}
