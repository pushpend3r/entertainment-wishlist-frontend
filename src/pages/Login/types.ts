import { Tokens } from "../../types";

export interface Login {
  login: Tokens;
}
