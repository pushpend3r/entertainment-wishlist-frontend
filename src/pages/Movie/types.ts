import { Movie } from "../../types";

export interface GetMovieDetails {
  getMovie: Movie;
}
