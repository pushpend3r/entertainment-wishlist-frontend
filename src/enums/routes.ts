enum ROUTES {
  HOME_PAGE = "/",

  WANNA_WATCH_PAGE = HOME_PAGE + "wanna-watch",
  ALREADY_WATCHED_PAGE = HOME_PAGE + "already-watched",

  REGISTER_PAGE = HOME_PAGE + "register",
  LOGIN_PAGE = HOME_PAGE + "login",
  PASSWORD_CHANGE_PAGE = HOME_PAGE + "password-change",

  MOVIE_PAGE = HOME_PAGE + "movie/:id",
  TVSHOW_PAGE = HOME_PAGE + "tvshow/:id",
}

export { ROUTES };
