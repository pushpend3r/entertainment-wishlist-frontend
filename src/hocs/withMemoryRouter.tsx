import { ComponentProps } from "react";
import { MemoryRouter, MemoryRouterProps } from "react-router-dom";

function withMemoryRouter<PropType>(
  Component: (props: PropType) => JSX.Element,
  initialEntries: MemoryRouterProps["initialEntries"] = ["/"]
) {
  return (props: ComponentProps<typeof Component> & JSX.IntrinsicAttributes) => {
    return (
      <MemoryRouter initialEntries={initialEntries}>
        <Component {...props} />
      </MemoryRouter>
    );
  };
}

export default withMemoryRouter;
