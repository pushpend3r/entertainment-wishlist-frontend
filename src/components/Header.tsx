import { useLocalStorage } from "@uidotdev/usehooks";
import Container from "react-bootstrap/Container";
import Nav from "react-bootstrap/Nav";
import Navbar from "react-bootstrap/Navbar";
import { NavLink, useLocation, useNavigate } from "react-router-dom";
import useIsUserLoggedIn from "../hooks/useIsUserLoggedIn";
import { useEffect } from "react";
import Button from "./Button";
import { ROUTES } from "../enums/routes";

export const LoggedInNavItems = [
  {
    to: ROUTES.WANNA_WATCH_PAGE,
    displayValue: "Wanna Watch",
  },
  {
    to: ROUTES.ALREADY_WATCHED_PAGE,
    displayValue: "Already Watched",
  },
  {
    to: ROUTES.PASSWORD_CHANGE_PAGE,
    displayValue: "Password Change",
  },
] as const;

export const LoggedOutNavItems = [
  {
    to: ROUTES.REGISTER_PAGE,
    displayValue: "Register",
  },
  {
    to: ROUTES.LOGIN_PAGE,
    displayValue: "Login",
  },
] as const;

function Header() {
  const navigate = useNavigate();
  const { pathname } = useLocation();

  const isUserLoggedIn = useIsUserLoggedIn();

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [_accessToken, setAccessToken] = useLocalStorage("accessToken");
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [_refreshToken, setRefreshToken] = useLocalStorage("refreshToken");

  const handleLogOutButtonClick = () => {
    setAccessToken("");
    setRefreshToken("");

    navigate(ROUTES.HOME_PAGE);
  };

  useEffect(() => {
    window.scrollTo(0, 0);
  }, [pathname]);

  return (
    <Navbar expand="lg" className="bg-body-tertiary">
      <Container>
        <Navbar.Brand as="div">
          <NavLink to={ROUTES.HOME_PAGE}>Entertainment Wishlist</NavLink>
        </Navbar.Brand>

        <Navbar.Toggle aria-controls="basic-navbar-nav" />

        <Navbar.Collapse id="basic-navbar-nav">
          <Nav className="ms-auto">
            {isUserLoggedIn ? (
              <>
                {LoggedInNavItems.map((item, index) => (
                  <Nav.Link as="div" key={index}>
                    <NavLink to={item.to}>{item.displayValue}</NavLink>
                  </Nav.Link>
                ))}
                <Button onClick={handleLogOutButtonClick}>Logout</Button>
              </>
            ) : (
              <>
                {LoggedOutNavItems.map((item, index) => (
                  <Nav.Link as="div" key={index}>
                    <NavLink to={item.to}>{item.displayValue}</NavLink>
                  </Nav.Link>
                ))}
              </>
            )}
          </Nav>
        </Navbar.Collapse>
      </Container>
    </Navbar>
  );
}

export default Header;
