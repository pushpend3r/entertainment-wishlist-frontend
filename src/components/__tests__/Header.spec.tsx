import { render, screen, within } from "@testing-library/react";
import Header, { LoggedInNavItems, LoggedOutNavItems } from "../Header";
import withMemoryRouter from "../../hocs/withMemoryRouter";
import { TokenType } from "../../types";
import { ROUTES } from "../../enums/routes";

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
window.scrollTo = vi.fn();

describe("<Header /> component", () => {
  afterEach(() => {
    vi.resetAllMocks();
  });

  afterAll(() => {
    vi.clearAllMocks();
  });

  test("user logged in snapshot match", () => {
    localStorage.setItem(TokenType.ACCESS_TOKEN, "123");

    const WithRouter = withMemoryRouter(Header);
    const { container } = render(<WithRouter />);
    expect(container).toMatchSnapshot();
  });

  test("logged in nav items render", () => {
    localStorage.setItem(TokenType.ACCESS_TOKEN, "123");

    const WithRouter = withMemoryRouter(Header);
    const { container } = render(<WithRouter />);
    LoggedInNavItems.forEach((item) => {
      const element = within(container).getByRole("link", { name: item.displayValue });
      expect(element).toHaveAttribute("href", item.to);
    });
  });

  test("user logged out snapshot match", () => {
    localStorage.removeItem(TokenType.ACCESS_TOKEN);

    const WithRouter = withMemoryRouter(Header);
    const { container } = render(<WithRouter />);
    expect(container).toMatchSnapshot();
  });

  test("logged out nav items render", () => {
    localStorage.removeItem(TokenType.ACCESS_TOKEN);

    const WithRouter = withMemoryRouter(Header);
    const { container } = render(<WithRouter />);

    LoggedOutNavItems.forEach((item) => {
      const element = within(container).getByRole("link", { name: item.displayValue });
      expect(element).toHaveAttribute("href", item.to);
    });
  });

  test("should have `/` link on logged in", () => {
    localStorage.setItem(TokenType.ACCESS_TOKEN, "123");

    const WithRouter = withMemoryRouter(Header);
    render(<WithRouter />);

    expect(screen.getByRole("link", { name: "Entertainment Wishlist" })).toHaveAttribute(
      "href",
      ROUTES.HOME_PAGE
    );
  });

  test("should have `/` link on logged out", () => {
    localStorage.removeItem(TokenType.ACCESS_TOKEN);

    const WithRouter = withMemoryRouter(Header);
    render(<WithRouter />);

    expect(screen.getByRole("link", { name: "Entertainment Wishlist" })).toHaveAttribute(
      "href",
      ROUTES.HOME_PAGE
    );
  });
});
