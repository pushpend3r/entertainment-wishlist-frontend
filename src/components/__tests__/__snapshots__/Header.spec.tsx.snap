// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Header /> component > user logged in snapshot match 1`] = `
<div>
  <nav
    class="bg-body-tertiary navbar navbar-expand-lg navbar-light"
  >
    <div
      class="container"
    >
      <div
        class="navbar-brand"
      >
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          Entertainment Wishlist
        </a>
      </div>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          class="ms-auto navbar-nav"
        >
          <div
            class="nav-link"
          >
            <a
              class=""
              href="/wanna-watch"
            >
              Wanna Watch
            </a>
          </div>
          <div
            class="nav-link"
          >
            <a
              class=""
              href="/already-watched"
            >
              Already Watched
            </a>
          </div>
          <div
            class="nav-link"
          >
            <a
              class=""
              href="/password-change"
            >
              Password Change
            </a>
          </div>
          <button
            class="btn btn-primary"
            type="button"
          >
            Logout
          </button>
        </div>
      </div>
    </div>
  </nav>
</div>
`;

exports[`<Header /> component > user logged out snapshot match 1`] = `
<div>
  <nav
    class="bg-body-tertiary navbar navbar-expand-lg navbar-light"
  >
    <div
      class="container"
    >
      <div
        class="navbar-brand"
      >
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          Entertainment Wishlist
        </a>
      </div>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          class="ms-auto navbar-nav"
        >
          <div
            class="nav-link"
          >
            <a
              class=""
              href="/register"
            >
              Register
            </a>
          </div>
          <div
            class="nav-link"
          >
            <a
              class=""
              href="/login"
            >
              Login
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
